### DISCUSSION TITLE
OpenSSL 3.x - How to use own signature function in TLS (convert engine to provider?)

### DISCUSSION THREAD

**Discussion by tanjahummel** [openssl/openssl#23117] [Q&A] [answered]
Hi, 
I have a similar question like this one https://github.com/openssl/openssl/discussions/22839
We are also developing C++ software that uses TLS with an own signature implementation (ECDSA) and are trying to convert the engine to a provider. As I already found out, I think we have to implement a provider for ec-keymgmt and signature.

Currently, the engine is set at the PKEY object which is passed to the SSL ctx like this:
EVP_PKEY* pub_key = X509_get_pubkey(certificate);
EVP_PKEY_set1_engine(pub_key, engine);
SSL_CTX_use_PrivateKey(ctx, pub_key);

1. I didn't found a similar way to configure the PKEY object to use my keymgmt and signature provider. How should this be done with the provider concept?

2. I want to use the default provider for everything except the sign function, like MscVZoltan also mentioned. Is there an easier way to do this, without having to implement all of the functions the default provider already implements?

**Comment by mattcaswell** [comment, marked as answer]
I answered in #22839, since the answer applies to both questions.
